Source/Core/DolphinQt/Config/LogWidget.h

```diff
@@ -43,12 +44,99 @@ class LogWidget final : public QDockWidget, LogListener
   QCheckBox* m_log_wrap;
   QComboBox* m_log_font;
   QPushButton* m_log_clear;
-  QVBoxLayout* m_main_layout;
-  QTextEdit* m_log_text;
-  QWidget* m_tab_log;
+  QPlainTextEdit* m_log_text;
 
   QTimer* m_timer;
 
+  // Ring buffer implementation, heavily based on:
+  // https://embeddedartistry.com/blog/2017/4/6/circular-buffers-in-cc#c-=
+  // Modified to be an almost complete drop-in replacement for std::queue
+  // It does not implement some methods compared to std::queue,
+  // and also implements capacity(), clear() and full(), which std::queue does not
+  template <class T>
+  class circular_buffer
+  {
+  public:
+    using value_type = T;
+    using size_type = std::size_t;
+    using reference = T&;
+    using const_reference = const T&;
+
+    explicit circular_buffer(std::size_t count)
+        : m_container(std::make_unique<storage[]>(count)), m_capacity(count)
+    {
+    }
```

Comment:
The destructor for this should really be invoking the destructor of the placed types should it not? Otherwise any type with a non-trivial destructor won't actually execute. If it exists within the buffer when its lifetime scope ends as far as I can tell, which is dangerous.

Reply:
Snap, you're absolutely right! I'll fix it, good catch!